firedrake/function.py

```diff
@@ -214,6 +177,125 @@ def __str__(self):
         else:
             return super(Function, self).__str__()
 
+    def as_coordinates(self):
+        """Create a mesh object using this function as coordinates."""
+        if hasattr(self, '_as_coordinates'):
+            mesh = self._as_coordinates()
+            if mesh is not None:
+                return mesh
+
+        from firedrake.mesh import MeshGeometry
+        return MeshGeometry(self)
```

Comment:
@dham has decided that we should *overload* the `Mesh` factory function, and we should add some checks (only from VectorFunctionSpace, affine mapping).